libbeat/beat/events/util.go

```diff
@@ -0,0 +1,70 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package events
+
+import "github.com/elastic/beats/v7/libbeat/beat"
+
+const (
+	// FieldMetaID defines the ID for the event. Also see FieldMetaOpType.
+	FieldMetaID = "_id"
+
+	// FieldMetaAlias defines the index alias to use for the event. If set, it takes
+	// precedence over values defined using FieldMetaIndex or FieldMetaRawIndex.
+	FieldMetaAlias = "alias"
+
+	// FieldMetaIndex defines the base index name to use for the event. The value is suffixed
+	// with a Y-m-d value based on the event's timestamp. If set, it takes precedence over the
+	// value defined using FieldMetaRawIndex.
+	FieldMetaIndex = "index"
+
+	// FieldMetaRawIndex defines the raw index name to use for the event. It is used as-is, without
+	// any additional manipulation.
+	FieldMetaRawIndex = "raw_index"
+
+	// FieldMetaPipeline defines the ingest node pipeline to use for this event.
+	FieldMetaPipeline = "pipeline"
+
+	// FieldMetaOpType defines the metadata key name for event operation type to use with the Elasticsearch
+	// Bulk API encoding of the event. The key's value can be an empty string, `create`, `index`, or `delete`.
+	// If empty, `create` will be used if FieldMetaID is set; otherwise `index` will be used.
+	FieldMetaOpType = "op_type"
+
+	FieldMetaOpTypeCreate MetaOpType = iota
+	FieldMetaOpTypeDelete
+	FieldMetaOpTypeIndex
+)
+
+type MetaOpType int
+
+func (o MetaOpType) String() string {
+	return []string{"create", "delete", "index"}[o]
+}
```

Comment:
an invalid value of `o` can create a panic. Using stringer you can do this:

```
type OpType int

//go:generate stringer -linecomment -type OpType
const (
  OpTypeCreate OpType = iota // create
  OpTypeIndex                // index
  OpTypeDelete               // delete
)
```

The create the `String` method via `go generate`. We use `go generate` and `stringer` in a few places already.
The generated `String` method will not panic if `o` is invalid.